src/AppBundle/Repository/EventCategoryRepository.php

```diff
@@ -26,57 +27,101 @@ public function getEntityClass()
+    public function getCategoryNameFromId($dbName, $categoryId)
     {
-        $ret = $this->getUsernamesFromIds([$categoryId]);
+        $ret = $this->getCategoryNamesFromIds($dbName, [$categoryId]);
         return isset($ret[0]['cat_title']) ? $ret[0]['cat_title'] : null;
     }
 
     /**
      * Fetch the ids of the categories with the given titles.
+     * @param string $dbName Database name such as 'enwiki_p'.
      * @param int[] $categoryNames
      * @return array with keys 'cat_title' and 'cat_id'.
      */
-    public function getCategoryIdsFromNames(array $categoryNames)
+    public function getCategoryIdsFromNames($dbName, array $categoryNames)
     {
         $rqb = $this->getReplicaConnection()->createQueryBuilder();
         $rqb->select(['cat_title', 'cat_id'])
-            ->from('category')
-            ->andWhere('cat_title IN (:titles)')
-            ->setParameter('titles', $categoryNames, Connection::PARAM_STR_ARRAY);
+            ->from("$dbName.category")
+            ->andWhere('cat_title IN (:categoryTitles)')
+            ->setParameter('categoryTitles', $categoryNames, Connection::PARAM_STR_ARRAY);
         // false means do not set a max query time. Here it's really fast,
         // and setting the query timeout actually slows it down.
         return $this->executeQueryBuilder($rqb, false)->fetchAll();
     }
 
     /**
      * Get the category ID given the name.
+     * @param string $dbName Database name such as 'enwiki_p'.
      * @param $categoryName
      * @return string|null Null if not found.
      */
-    public function getCategoryIdFromName($categoryName)
+    public function getCategoryIdFromName($dbName, $categoryName)
     {
-        $ret = $this->getUsernamesFromIds([$categoryName]);
+        $ret = $this->getCategoryIdsFromNames($dbName, [$categoryName]);
         return isset($ret[0]['cat_id']) ? $ret[0]['cat_id'] : null;
     }
+
+    /**
+     * Get the IDs of pages in the given categories.
+     * @param string $dbName Database name such as 'enwiki_p'.
+     * @param array $categoryIds IDs of categories to fetch from.
+     * @param bool $queryBuilder Whether to return just the Doctrine query builder object.
+     * @param int|null $limit Max number of pages. null for no limit, but only do this if used in a subquery.
+     * @return array|QueryBuilder Page IDs or the QueryBuilder object.
+     */
+    public function getPagesInCategories($dbName, array $categoryIds, $queryBuilder = false, $limit = 20000)
+    {
+        $rqb = $this->getReplicaConnection()->createQueryBuilder();
+        $rqb->select(['DISTINCT(cl_from)'])
+            ->from("$dbName.categorylinks")
+            ->join("$dbName.categorylinks", "$dbName.page", null, 'cl_from = page_id')
+            ->where('page_namespace = 0')
+            ->andWhere(
+                $rqb->expr()->in(
+                    'cl_to',
+                    "SELECT cat_title FROM $dbName.category WHERE cat_id IN (:categoryIds)"
```

Comment:
I'm probably not thinking about this correctly, but do we need to do this with a subquery? Can we join categorylinks to both category and page, e.g.

```php
    $rqb->select(['DISTINCT(cl_from)'])
            ->from("$dbName.categorylinks")
            ->join("$dbName.categorylinks", "$dbName.category", 'cl', 'cl_to = cat_title')
            ->join("$dbName.categorylinks", "$dbName.page", 'p', 'cl_from = page_id')
            ->where('page_namespace = 0')
            ->where( 'cat_id IN (:categoryIds)'
```

Reply:
I believe you are correct!